File: T-Bank/Algorithms-and-Data-Structures-2025/12sprint-SolvingProblemMethodsPart2/contest/A-LCA/solution.cpp
#include <iostream>
#include <vector>

void dfs(
    int v,
    std::vector<std::vector<int>> &adjList,
    std::vector<std::vector<int>> &dp,
    std::vector<int> &timeIn,
    std::vector<int> &timeOut,
    int &t
) {
    timeIn[v] = t++;
    for (int u : adjList[v]) {
        dp[u][0] = v;
        dfs(u, adjList, dp, timeIn, timeOut, t);
    }
    timeOut[v] = t++;
}

bool isAncestor(int u, int v, std::vector<int> &timeIn, std::vector<int> &timeOut) {
    return timeIn[u] <= timeIn[v] && timeOut[u] >= timeOut[v];
}

int lca(
    int u, int v,
    std::vector<std::vector<int>> &dp,
    std::vector<int> &timeIn,
    std::vector<int> &timeOut
) {
    if (isAncestor(u, v, timeIn, timeOut)) return u;
    if (isAncestor(v, u, timeIn, timeOut)) return v;
    for (int level = 30; level >= 0; --level) {
        if (!isAncestor(dp[u][level], v, timeIn, timeOut)) {
            u = dp[u][level];
        }
    }
    return dp[u][0];
}

void solution() {
    /*
    Time Complexity O(NlogN + MlogN)
    Memory Complexity O(N)
    */
    int n;
    std::scanf("%d", &n);
    std::vector<std::vector<int>> adjList(n);
    std::vector<std::vector<int>> dp(n, std::vector<int>(31, 0));
    for (int i = 1; i < n; ++i) {
        int parent;
        std::scanf("%d", &parent);
        adjList[parent].push_back(i);
    }
    std::vector<int> timeIn(n), timeOut(n);
    int t = 0;
    dfs(0, adjList, dp, timeIn, timeOut, t);
    for (int v = 0; v < n; ++v) {
        for (int level = 1; level < 31; ++level) {
            dp[v][level] = dp[dp[v][level - 1]][level - 1];
        }
    }
    int m;
    std::scanf("%d", &m);
    for (int i = 0; i < m; ++i) {
        int u, v;
        std::scanf("%d %d", &u, &v);
        std::printf("%d\n", lca(u, v, dp, timeIn, timeOut));
    }
}

int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/12sprint-SolvingProblemMethodsPart2/contest/A-LCA/solution.py
import sys


def dfs(v: int, adj_list: list[list[int]], dp: list[list[int]], time_in: list[int], time_out: list[int], t: list[int]) -> None:
    time_in[v] = t[0]
    t[0] += 1
    for u in adj_list[v]:
        dp[u][0] = v
        dfs(u, adj_list, dp, time_in, time_out, t)
    time_out[v] = t[0]
    t[0] += 1


def is_ancestor(u: int, v: int, time_in: list[int], time_out: list[int]) -> bool:
    return time_in[u] <= time_in[v] and time_out[v] <= time_out[u]


def lca(u: int, v: int, dp: list[list[int]], time_in: list[int], time_out: list[int]) -> int:
    if is_ancestor(u, v, time_in, time_out): return u
    elif is_ancestor(v, u, time_in, time_out): return v
    for level in range(30, -1, -1):
        if not is_ancestor(dp[u][level], v, time_in, time_out):
            u = dp[u][level]
    return dp[u][0]


def solution() -> None:
    """
    Time Complexity O(NlogN + MlogN)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    adj_list: list[list[int]] = [[] for _ in range(n)]
    vertices: iter[int] = map(int, sys.stdin.readline().rstrip().split())
    for i in range(1, n):
        parent: int = next(vertices)
        adj_list[parent].append(i)
    time_in: list[int] = [0] * n
    time_out: list[int] = [0] * n
    dp: list[list[int]] = [[0] * 31 for _ in range(n)]
    dfs(0, adj_list, dp, time_in, time_out, [0])
    for v in range(n):
        for level in range(1, 31):
            dp[v][level] = dp[dp[v][level - 1]][level - 1]
    m: int = int(sys.stdin.readline().rstrip())
    for _ in range(m):
        u, v = map(int, sys.stdin.readline().rstrip().split())
        sys.stdout.write('{}\n'.format(lca(u, v, dp, time_in, time_out)))
    


if __name__ == '__main__':
    solution()